File: java8/src/main/java/com/shekhargulati/ninetynine_problems/_01_lists/P03.java
package com.shekhargulati.ninetynine_problems._01_lists;

import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Find the Kth element of a list.
 */
public class P03 {

    public static <T> T kth(final List<T> list, final int k) {
        return list.get(k);
    }

    public static <T> T kthRecursive(final LinkedList<T> list, final int k) {
        if (k == 0) {
            return list.getFirst();
        }
        return kthRecursive(new LinkedList<>(list.subList(1, list.size())), k - 1);
    }

    public static <T> T kthStream(final List<T> list, final int k) {
        return list.stream().limit(k + 1).collect(Collectors.toCollection(LinkedList::new)).getLast();
    }

}

File: scala/src/main/scala/com/shekhargulati/ninetynine_problems/_01_lists/P03.scala
package com.shekhargulati.ninetynine_problems._01_lists

/**
  * Find the K'th element of a list.
  *
  * The first element of a list has index 0
  */
object P03 {

  def kth[T](list: List[T], k: Int): T = list(k)

  def kth1[T](list: List[T], k: Int): T = list.take(k + 1).last

  def kthRecursive[T](list: List[T], k: Int): T = list match {
    case x :: xs if k == 0 => x
    case x :: xs => kthRecursive(xs, k - 1)
    case _ => throw new NoSuchElementException
  }


}
